From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Glitch <glitch.g3431@gmail.com>
Date: Sat, 12 Mar 2022 20:35:46 -0600
Subject: [PATCH] Fix buildscript


diff --git a/build.gradle b/build.gradle
index 9e35f2d655d34f904ea54d2c6ea5de794bf2807c..15ceecae59d19502cb20e12c82dfed6c15183a1a 100644
--- a/build.gradle
+++ b/build.gradle
@@ -290,9 +290,6 @@ task writeActionsTestMatrix() {
 	}
 }
 
-tasks.named('wrapper') {
-	distributionType = Wrapper.DistributionType.ALL
-}
 
 /**
  * Run this task to download the gradle sources next to the api jar, you may need to manually attach the sources jar
