From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Glitch <glitch.g3431@gmail.com>
Date: Thu, 9 Jun 2022 23:40:54 -0500
Subject: [PATCH] Exclude Fabric dependencies replaced by Quilt


diff --git a/src/main/java/net/fabricmc/loom/configuration/InstallerData.java b/src/main/java/net/fabricmc/loom/configuration/InstallerData.java
index ce8da78d..0d77a024 100644
--- a/src/main/java/net/fabricmc/loom/configuration/InstallerData.java
+++ b/src/main/java/net/fabricmc/loom/configuration/InstallerData.java
@@ -40,9 +40,8 @@ import net.fabricmc.loom.LoomRepositoryPlugin;
 import net.fabricmc.loom.configuration.ide.idea.IdeaUtils;
 import net.fabricmc.loom.util.Constants;
 
-public record InstallerData(String version, JsonObject installerJson) {
+public record InstallerData(String version, JsonObject installerJson, boolean isQuilt) {
 	private static final Logger LOGGER = LoggerFactory.getLogger(InstallerData.class);
-
 	public void applyToProject(Project project) {
 		LoomGradleExtension extension = LoomGradleExtension.get(project);
 
diff --git a/src/main/java/net/fabricmc/loom/configuration/mods/ArtifactMetadata.java b/src/main/java/net/fabricmc/loom/configuration/mods/ArtifactMetadata.java
index c7e91aca..c1665d6e 100644
--- a/src/main/java/net/fabricmc/loom/configuration/mods/ArtifactMetadata.java
+++ b/src/main/java/net/fabricmc/loom/configuration/mods/ArtifactMetadata.java
@@ -44,7 +44,8 @@ import net.fabricmc.loom.util.FileSystemUtil;
 import net.fabricmc.loom.util.fmj.FabricModJsonFactory;
 
 public record ArtifactMetadata(boolean isFabricMod, RemapRequirements remapRequirements, @Nullable InstallerData installerData, MixinRemapType mixinRemapType) {
-	private static final String INSTALLER_PATH = "quilt_installer.json";
+	private static final String QUILT_INSTALLER_PATH = "quilt_installer.json";
+	private static final String FABRIC_INSTALLER_PATH = "fabric-installer.json";
 
 	public static ArtifactMetadata create(ArtifactRef artifact, String currentLoomVersion) throws IOException {
 		boolean isFabricMod;
@@ -81,11 +82,17 @@ public record ArtifactMetadata(boolean isFabricMod, RemapRequirements remapRequi
 				}
 			}
 
-			final Path installerPath = fs.getPath(INSTALLER_PATH);
+			Path installerPath = fs.getPath(QUILT_INSTALLER_PATH);
+			boolean isQuilt = true;
+
+			if (!Files.exists(installerPath)) {
+				installerPath = fs.getPath(FABRIC_INSTALLER_PATH);
+				isQuilt = false;
+			}
 
 			if (isFabricMod && Files.exists(installerPath)) {
 				final JsonObject jsonObject = LoomGradlePlugin.GSON.fromJson(Files.readString(installerPath, StandardCharsets.UTF_8), JsonObject.class);
-				installerData = new InstallerData(artifact.version(), jsonObject);
+				installerData = new InstallerData(artifact.version(), jsonObject, isQuilt);
 			}
 		}
 
diff --git a/src/main/java/net/fabricmc/loom/configuration/mods/ModConfigurationRemapper.java b/src/main/java/net/fabricmc/loom/configuration/mods/ModConfigurationRemapper.java
index 61a4d844..2cd92134 100644
--- a/src/main/java/net/fabricmc/loom/configuration/mods/ModConfigurationRemapper.java
+++ b/src/main/java/net/fabricmc/loom/configuration/mods/ModConfigurationRemapper.java
@@ -32,12 +32,20 @@ import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.Comparator;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Supplier;
 
 import com.google.common.collect.ImmutableMap;
+
+import net.fabricmc.loom.configuration.InstallerData;
+
+import net.fabricmc.loom.configuration.mods.dependency.ModDependencyFactory;
+
 import org.gradle.api.Project;
 import org.gradle.api.artifacts.Configuration;
 import org.gradle.api.artifacts.FileCollectionDependency;
@@ -61,7 +69,6 @@ import net.fabricmc.loom.LoomGradlePlugin;
 import net.fabricmc.loom.api.RemapConfigurationSettings;
 import net.fabricmc.loom.configuration.RemapConfigurations;
 import net.fabricmc.loom.configuration.mods.dependency.ModDependency;
-import net.fabricmc.loom.configuration.mods.dependency.ModDependencyFactory;
 import net.fabricmc.loom.configuration.providers.minecraft.MinecraftSourceSets;
 import net.fabricmc.loom.util.Checksum;
 import net.fabricmc.loom.util.Constants;
@@ -133,20 +140,21 @@ public class ModConfigurationRemapper {
 		}
 
 		// Round 1: Discovery
-		// Go through all the configs to find artifacts to remap and
-		// the installer data. The installer data has to be added before
+		// Go through all the configs and find the installer data and QSL.
+		// The installer data has to be added before
 		// any mods are remapped since remapping needs the dependencies provided by that data.
 		final Map<Configuration, List<ModDependency>> dependenciesBySourceConfig = new HashMap<>();
+		final Map<Configuration, List<ArtifactRef>> artifactsBySourceConfig = new HashMap<>();
+		final Set<Configuration> sourceConfigsWithQsl = new HashSet<>();
+		// todo clean this up
+		AtomicReference<ArtifactMetadata> installer = new AtomicReference<>();
+		AtomicReference<Configuration> installerConfigToApply = new AtomicReference<>();
+		AtomicReference<ArtifactRef> installerArtifactToApply = new AtomicReference<>();
 		final Map<ArtifactRef, ArtifactMetadata> metaCache = new HashMap<>();
 		configsToRemap.forEach((sourceConfig, remappedConfig) -> {
-			/*
-			sourceConfig - The source configuration where the intermediary named artifacts come from. i.e "modApi"
-			remappedConfig - The target configuration where the remapped artifacts go
-			 */
-			final Configuration clientRemappedConfig = clientConfigsToRemap.get(sourceConfig);
-			final List<ModDependency> modDependencies = new ArrayList<>();
+			List<ArtifactRef> artifacts = resolveArtifacts(project, sourceConfig);
 
-			for (ArtifactRef artifact : resolveArtifacts(project, sourceConfig)) {
+			for (ArtifactRef artifact : artifacts) {
 				final ArtifactMetadata artifactMetadata;
 
 				artifactMetadata = metaCache.computeIfAbsent(artifact, a -> {
@@ -158,21 +166,84 @@ public class ModConfigurationRemapper {
 				});
 
 				if (artifactMetadata.installerData() != null) {
-					if (extension.getInstallerData() != null) {
-						project.getLogger().info("Found another installer JSON in ({}), ignoring", artifact.path());
+					if (installer.get() == null) {
+						installer.set(artifactMetadata);
+					} else if (!installer.get().installerData().isQuilt() && artifactMetadata.installerData().isQuilt()) {
+						// Prefer Quilt over Fabric loader
+						installer.set(artifactMetadata);
 					} else {
-						project.getLogger().info("Applying installer data from {}", artifact.path());
-						artifactMetadata.installerData().applyToProject(project);
+						project.getLogger().info("Found another installer JSON in ({}), ignoring", artifact.path());
 					}
 				}
 
 				if (!artifactMetadata.shouldRemap()) {
+					if (artifactMetadata.installerData() != null) {
+						installerConfigToApply.set(remappedConfig);
+						installerArtifactToApply.set(artifact);
+						continue; // we'll apply the installer later
+					}
 					// Note: not applying to any type of vanilla Gradle target config like
 					// api or implementation to fix https://github.com/FabricMC/fabric-loom/issues/572.
 					artifact.applyToConfiguration(project, remappedConfig);
 					continue;
 				}
 
+				if (artifact.group().startsWith("org.quiltmc.qsl")) {
+					sourceConfigsWithQsl.add(sourceConfig);
+				}
+			}
+
+			artifactsBySourceConfig.put(sourceConfig, artifacts);
+		});
+
+		boolean quiltLoader;
+
+		// Apply the installer.
+		// We do this outside Round 1, unlike Fabric Loom, because we need to prioritize Quilt Loader over Fabric Loader.
+		if (installer.get() != null) {
+			project.getLogger().info("Applying installer data");
+
+            if (installer.get().installerData().isQuilt()) {
+                quiltLoader = true;
+            } else {
+				quiltLoader = false;
+				project.getLogger().lifecycle("Warning: Quilt Loader not detected. Using Fabric Loader.");
+            }
+
+            if (!installer.get().shouldRemap()) {
+				installerArtifactToApply.get().applyToConfiguration(project, installerConfigToApply.get());
+			}
+			installer.get().installerData().applyToProject(project);
+		} else {
+			project.getLogger().lifecycle("Warning: No loader detected.");
+			quiltLoader = false;
+		}
+
+		// Round 1.5: Proposal
+		// Propose everything to be remapped, excluding anything we want to hide.
+		configsToRemap.forEach((sourceConfig, remappedConfig) -> {
+			/*
+			sourceConfig - The source configuration where the intermediary named artifacts come from. i.e "modApi"
+			remappedConfig - The target configuration where the remapped artifacts go
+			 */
+			final Configuration clientRemappedConfig = clientConfigsToRemap.get(sourceConfig);
+			final List<ModDependency> modDependencies = new ArrayList<>();
+
+			for (ArtifactRef artifact : artifactsBySourceConfig.get(sourceConfig)) {
+				if (quiltLoader && artifact.group().equals("net.fabricmc") && artifact.name().equals("fabric-loader")) {
+					continue;
+				} else if (sourceConfigsWithQsl.contains(sourceConfig) && artifact.group().equals("net.fabricmc.fabric-api")) {
+					continue;
+				}
+				final ArtifactMetadata artifactMetadata;
+
+				artifactMetadata = metaCache.computeIfAbsent(artifact, a -> {
+					try {
+						return ArtifactMetadata.create(a, LoomGradlePlugin.LOOM_VERSION);
+					} catch (IOException e) {
+						throw ExceptionUtil.createDescriptiveWrapper(UncheckedIOException::new, "Failed to read metadata from " + a.path(), e);
+					}
+				});
 				final ModDependency modDependency = ModDependencyFactory.create(artifact, artifactMetadata, remappedConfig, clientRemappedConfig, mappingsSuffix, project);
 				scheduleSourcesRemapping(project, sourceRemapper, modDependency);
 				modDependencies.add(modDependency);
diff --git a/src/test/groovy/net/fabricmc/loom/test/integration/MavenProjectTest.groovy b/src/test/groovy/net/fabricmc/loom/test/integration/MavenProjectTest.groovy
index 1a546c22..b56f85b7 100644
--- a/src/test/groovy/net/fabricmc/loom/test/integration/MavenProjectTest.groovy
+++ b/src/test/groovy/net/fabricmc/loom/test/integration/MavenProjectTest.groovy
@@ -46,6 +46,7 @@ class MavenProjectTest extends Specification implements MockMavenServerTrait, Gr
 	def "publish lib #version #gradleVersion"() {
 		setup:
 		setProperty('loom.test.version', version)
+		setProperty('loom.excludeFabricReplacedDependencies', 'false')
 		def gradle = gradleProject(project: "mavenLibrary", version: gradleVersion, sharedFiles: true)
 
 		when:
@@ -73,6 +74,7 @@ class MavenProjectTest extends Specification implements MockMavenServerTrait, Gr
 	def "resolve #version #gradleVersion"() {
 		given:
 		setProperty('loom.test.resolve', "com.example:fabric-example-lib:${version}")
+		setProperty('loom.excludeFabricReplacedDependencies', 'false')
 		def gradle = gradleProject(project: "maven", version: gradleVersion, sharedFiles: true)
 
 		when:
diff --git a/src/test/groovy/net/fabricmc/loom/test/integration/MultiProjectTest.groovy b/src/test/groovy/net/fabricmc/loom/test/integration/MultiProjectTest.groovy
index c5669df0..18e060e5 100644
--- a/src/test/groovy/net/fabricmc/loom/test/integration/MultiProjectTest.groovy
+++ b/src/test/groovy/net/fabricmc/loom/test/integration/MultiProjectTest.groovy
@@ -29,6 +29,7 @@ import spock.lang.Unroll
 
 import net.fabricmc.loom.test.util.GradleProjectTestTrait
 
+import static java.lang.System.setProperty
 import static net.fabricmc.loom.test.LoomTestConstants.STANDARD_TEST_VERSIONS
 import static org.gradle.testkit.runner.TaskOutcome.SUCCESS
 
@@ -36,6 +37,7 @@ class MultiProjectTest extends Specification implements GradleProjectTestTrait {
 	@Unroll
 	def "build (gradle #version)"() {
 		setup:
+		setProperty('loom.excludeFabricReplacedDependencies', 'false')
 		def gradle = gradleProject(project: "multiproject", version: version)
 
 		when:
