From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ByMartrixx <bymartrixx@gmail.com>
Date: Mon, 30 May 2022 14:16:47 -0500
Subject: [PATCH] Fix mappingsConstants dependency being added even if already
 present

Implements https://github.com/FabricMC/fabric-loom/pull/605

diff --git a/src/main/java/net/fabricmc/loom/configuration/providers/mappings/MappingConfiguration.java b/src/main/java/net/fabricmc/loom/configuration/providers/mappings/MappingConfiguration.java
index 5cf477755e4666e91c86a27f361f22653ab405a1..c775059042d32d5347d6a794a97b90508924c22e 100644
--- a/src/main/java/net/fabricmc/loom/configuration/providers/mappings/MappingConfiguration.java
+++ b/src/main/java/net/fabricmc/loom/configuration/providers/mappings/MappingConfiguration.java
@@ -141,13 +141,19 @@ public class MappingConfiguration {
 
 	public void applyToProject(Project project, DependencyInfo dependency) {
 		if (hasUnpickDefinitions()) {
-			String notation = String.format("%s:%s:%s:constants",
-					dependency.getDependency().getGroup(),
-					dependency.getDependency().getName(),
-					dependency.getDependency().getVersion()
-			);
+			boolean layered = dependency.getDependency() instanceof LayeredMappingsDependency;
+
+			if (!layered && project.getConfigurations().getByName(Constants.Configurations.MAPPING_CONSTANTS).getDependencies().isEmpty()) {
+				// Add a constants jar dependency if not already present
+				String notation = String.format("%s:%s:%s:constants",
+						dependency.getDependency().getGroup(),
+						dependency.getDependency().getName(),
+						dependency.getDependency().getVersion()
+				);
+
+				project.getDependencies().add(Constants.Configurations.MAPPING_CONSTANTS, notation);
+			}
 
-			project.getDependencies().add(Constants.Configurations.MAPPING_CONSTANTS, notation);
 			populateUnpickClasspath(project);
 		}
 
