From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ByMartrixx <bymartrixx@gmail.com>
Date: Mon, 30 May 2022 14:16:47 -0500
Subject: [PATCH] Fix mappingsConstants dependency being added even if already
 present

Implements https://github.com/FabricMC/fabric-loom/pull/605

diff --git a/src/main/java/net/fabricmc/loom/configuration/providers/mappings/MappingsProviderImpl.java b/src/main/java/net/fabricmc/loom/configuration/providers/mappings/MappingsProviderImpl.java
index c780a8cc63ced2e7dba8704104e7a0b5a069e727..e14c160c61bf3b4fa01d41759e9e3f44e4cff22d 100644
--- a/src/main/java/net/fabricmc/loom/configuration/providers/mappings/MappingsProviderImpl.java
+++ b/src/main/java/net/fabricmc/loom/configuration/providers/mappings/MappingsProviderImpl.java
@@ -159,13 +159,19 @@ public class MappingsProviderImpl implements MappingsProvider, SharedService {
 
 	public void applyToProject(Project project, DependencyInfo dependency) {
 		if (hasUnpickDefinitions()) {
-			String notation = String.format("%s:%s:%s:constants",
-					dependency.getDependency().getGroup(),
-					dependency.getDependency().getName(),
-					dependency.getDependency().getVersion()
-			);
+			boolean layered = dependency.getDependency() instanceof LayeredMappingsDependency;
+
+			if (!layered && project.getConfigurations().getByName(Constants.Configurations.MAPPING_CONSTANTS).getDependencies().isEmpty()) {
+				// Add a constants jar dependency if not already present
+				String notation = String.format("%s:%s:%s:constants",
+						dependency.getDependency().getGroup(),
+						dependency.getDependency().getName(),
+						dependency.getDependency().getVersion()
+				);
+
+				project.getDependencies().add(Constants.Configurations.MAPPING_CONSTANTS, notation);
+			}
 
-			project.getDependencies().add(Constants.Configurations.MAPPING_CONSTANTS, notation);
 			populateUnpickClasspath(project);
 		}
 
