From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Luke Bemish <lukebemish@lukebemish.dev>
Date: Sun, 31 Mar 2024 19:32:07 -0500
Subject: [PATCH] Exclude quilt-loader dependencies from remapped
 configurations


diff --git a/src/main/java/net/fabricmc/loom/configuration/RemapConfigurations.java b/src/main/java/net/fabricmc/loom/configuration/RemapConfigurations.java
index e0607b33f52c5ded85a67c37012d0f0c8cc588d1..89fda57d069f66cfd4928478421057515b0031a7 100644
--- a/src/main/java/net/fabricmc/loom/configuration/RemapConfigurations.java
+++ b/src/main/java/net/fabricmc/loom/configuration/RemapConfigurations.java
@@ -26,6 +26,7 @@ package net.fabricmc.loom.configuration;
 
 import java.util.List;
 import java.util.Locale;
+import java.util.Map;
 import java.util.function.Function;
 
 import org.gradle.api.Action;
@@ -177,6 +178,8 @@ public final class RemapConfigurations {
 			extendsFrom(Constants.Configurations.MOD_COMPILE_CLASSPATH, configuration, project);
 		}
 
+		configuration.exclude(Map.of("group", "org.quiltmc", "module", "quilt-loader-dependencies"));
+
 		for (String outgoingConfigurationName : settings.getPublishingMode().get().outgoingConfigurations()) {
 			extendsFrom(outgoingConfigurationName, configuration, project);
 		}
