From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ennui Langeweile <85590273+EnnuiL@users.noreply.github.com>
Date: Mon, 7 Oct 2024 03:12:42 -0300
Subject: [PATCH] Support QMJ5


diff --git a/build.gradle b/build.gradle
index 9ecb7aed2f3eaf8c86bb1eb8fab4597d859e58d3..5f7fac291e66298fbbb7152cfbb47644544ce440 100644
--- a/build.gradle
+++ b/build.gradle
@@ -51,6 +51,10 @@ repositories {
 		name = 'Fabric'
 		url = 'https://maven.fabricmc.net/'
 	}
+	maven {
+		name = 'Quilt'
+		url = 'https://maven.quiltmc.org/repository/release'
+	}
 	mavenCentral()
 }
 
@@ -95,6 +99,7 @@ dependencies {
 	// libraries
 	implementation libs.gson
 	implementation libs.bundles.asm
+	implementation libs.bundles.quilt.parsers
 
 	// game handling utils
 	implementation (libs.fabric.stitch) {
diff --git a/gradle/libs.versions.toml b/gradle/libs.versions.toml
index 8c02b64a7fd1c827ec0e972f80deb2b48bb349b8..cb31bf0a1ce02d2c0a3cf3821d1134d2865eb352 100644
--- a/gradle/libs.versions.toml
+++ b/gradle/libs.versions.toml
@@ -2,6 +2,7 @@
 kotlin = "2.2.20"
 asm = "9.9"
 gson = "2.10.1"
+quilt-parsers = "0.3.0"
 
 stitch = "0.6.2"
 tiny-remapper = "0.12.0"
@@ -29,6 +30,9 @@ asm-util = { module = "org.ow2.asm:asm-util", version.ref = "asm" }
 
 gson = { module = "com.google.code.gson:gson", version.ref = "gson" }
 
+quilt-parsers-gson = { module = "org.quiltmc.parsers:gson", version.ref = "quilt-parsers" }
+quilt-parsers-json = { module = "org.quiltmc.parsers:json", version.ref = "quilt-parsers" }
+
 fabric-stitch = { module = "net.fabricmc:stitch", version.ref = "stitch" }
 fabric-tiny-remapper = { module = "net.fabricmc:tiny-remapper", version.ref = "tiny-remapper" }
 fabric-clazz-tweaker = { module = "net.fabricmc:class-tweaker", version.ref = "clazz-tweaker" }
@@ -51,3 +55,4 @@ retry = { id = "org.gradle.test-retry", version.ref = "test-retry" }
 
 [bundles]
 asm = ["asm", "asm-analysis", "asm-commons", "asm-tree", "asm-util"]
+quilt-parsers = ["quilt-parsers-gson", "quilt-parsers-json"]
diff --git a/src/main/java/net/fabricmc/loom/task/AbstractRemapJarTask.java b/src/main/java/net/fabricmc/loom/task/AbstractRemapJarTask.java
index 4ec34c3137d32f47819a85ac07db91014bc54067..755e901a91f2f9c1f68006ebc02feab2da52cafb 100644
--- a/src/main/java/net/fabricmc/loom/task/AbstractRemapJarTask.java
+++ b/src/main/java/net/fabricmc/loom/task/AbstractRemapJarTask.java
@@ -27,6 +27,7 @@ package net.fabricmc.loom.task;
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.StandardCopyOption;
@@ -39,6 +40,7 @@ import java.util.jar.Manifest;
 
 import javax.inject.Inject;
 
+import com.google.gson.JsonElement;
 import org.gradle.api.Action;
 import org.gradle.api.file.ConfigurableFileCollection;
 import org.gradle.api.file.RegularFileProperty;
@@ -61,8 +63,11 @@ import org.gradle.workers.WorkerExecutor;
 import org.jetbrains.annotations.ApiStatus;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import org.quiltmc.parsers.json.JsonReader;
+import org.quiltmc.parsers.json.gson.GsonReader;
 
 import net.fabricmc.loom.LoomGradleExtension;
+import net.fabricmc.loom.LoomGradlePlugin;
 import net.fabricmc.loom.api.mappings.layered.MappingsNamespace;
 import net.fabricmc.loom.task.service.ClientEntriesService;
 import net.fabricmc.loom.task.service.JarManifestService;
@@ -71,6 +76,7 @@ import net.fabricmc.loom.util.Constants;
 import net.fabricmc.loom.util.ExceptionUtil;
 import net.fabricmc.loom.util.ZipReprocessorUtil;
 import net.fabricmc.loom.util.ZipUtils;
+import net.fabricmc.loom.util.fmj.FabricModJsonFactory;
 import net.fabricmc.loom.util.gradle.SourceSetHelper;
 import net.fabricmc.loom.util.service.ScopedServiceFactory;
 
@@ -277,6 +283,20 @@ public abstract class AbstractRemapJarTask extends Jar {
 			Check.require(count > 0, "Did not transform any jar manifest");
 		}
 
+		protected void processQmj5sIntoQmjs() throws IOException {
+			if (ZipUtils.contains(outputFile, FabricModJsonFactory.QUILT_MOD_JSON5)) {
+				ZipUtils.transform(outputFile, Map.of(FabricModJsonFactory.QUILT_MOD_JSON5, bytes -> {
+					var jsonReader = new GsonReader(JsonReader.json5(new String(bytes, StandardCharsets.UTF_8)));
+					var element = LoomGradlePlugin.GSON.fromJson(jsonReader, JsonElement.class);
+					var convertedElement = LoomGradlePlugin.GSON.toJson(element);
+
+					return convertedElement.getBytes(StandardCharsets.UTF_8);
+				}));
+
+				ZipUtils.move(outputFile, FabricModJsonFactory.QUILT_MOD_JSON5, FabricModJsonFactory.QUILT_MOD_JSON);
+			}
+		}
+
 		protected void rewriteJar() throws IOException {
 			final boolean isReproducibleFileOrder = getParameters().getArchiveReproducibleFileOrder().get();
 			final boolean isPreserveFileTimestamps = getParameters().getArchivePreserveFileTimestamps().get();
diff --git a/src/main/java/net/fabricmc/loom/task/RemapJarTask.java b/src/main/java/net/fabricmc/loom/task/RemapJarTask.java
index f2cc03b0594126b6d5016fc6a4d5897f31530bea..ef5d2343b5c2ef316ffc4911882acd58a4aacbed 100644
--- a/src/main/java/net/fabricmc/loom/task/RemapJarTask.java
+++ b/src/main/java/net/fabricmc/loom/task/RemapJarTask.java
@@ -202,6 +202,7 @@ public abstract class RemapJarTask extends AbstractRemapJarTask {
 				addRefmaps(serviceFactory);
 				addNestedJars();
 				modifyJarManifest();
+				processQmj5sIntoQmjs();
 				rewriteJar();
 
 				if (getParameters().getOptimizeFmj().get()) {
diff --git a/src/main/java/net/fabricmc/loom/task/RemapSourcesJarTask.java b/src/main/java/net/fabricmc/loom/task/RemapSourcesJarTask.java
index 4a92c257ab7262db2b3732b3d9c9dd8f4b0c104c..c715478b57e278659b2f063edbcc3935b2105f2d 100644
--- a/src/main/java/net/fabricmc/loom/task/RemapSourcesJarTask.java
+++ b/src/main/java/net/fabricmc/loom/task/RemapSourcesJarTask.java
@@ -91,6 +91,7 @@ public abstract class RemapSourcesJarTask extends AbstractRemapJarTask {
 			}
 
 			modifyJarManifest();
+			processQmj5sIntoQmjs();
 			rewriteJar();
 		}
 	}
diff --git a/src/main/java/net/fabricmc/loom/task/launch/GenerateDLIConfigTask.java b/src/main/java/net/fabricmc/loom/task/launch/GenerateDLIConfigTask.java
index 0fae8cb4f3c3685f31038bc2034c93c3518230c5..aee9f30acde33672aebf405c7825fa89c02208cc 100644
--- a/src/main/java/net/fabricmc/loom/task/launch/GenerateDLIConfigTask.java
+++ b/src/main/java/net/fabricmc/loom/task/launch/GenerateDLIConfigTask.java
@@ -133,6 +133,7 @@ public abstract class GenerateDLIConfigTask extends AbstractLoomTask {
 
 		final LaunchConfig launchConfig = new LaunchConfig()
 				.property("loader.development", "true")
+				.property("loader.enable_quilt_mod_json5_in_dev_env", "true")
 				.property("log4j.configurationFile", getLog4jConfigPaths().get())
 				.property("log4j2.formatMsgNoLookups", "true")
 				.property("loader.defaultModDistributionNamespace", getProductionNamespace().get())
diff --git a/src/main/java/net/fabricmc/loom/util/ZipUtils.java b/src/main/java/net/fabricmc/loom/util/ZipUtils.java
index 29884c8b3e67535e22e00b377aa817676fc22153..a1a0962c05114a9ef786c0880b03788a1e15e2b5 100644
--- a/src/main/java/net/fabricmc/loom/util/ZipUtils.java
+++ b/src/main/java/net/fabricmc/loom/util/ZipUtils.java
@@ -47,6 +47,8 @@ import org.jspecify.annotations.Nullable;
 import org.objectweb.asm.ClassReader;
 import org.objectweb.asm.ClassVisitor;
 import org.objectweb.asm.ClassWriter;
+import org.quiltmc.parsers.json.JsonReader;
+import org.quiltmc.parsers.json.gson.GsonReader;
 
 import net.fabricmc.loom.LoomGradlePlugin;
 
@@ -112,6 +114,12 @@ public class ZipUtils {
 		return LoomGradlePlugin.GSON.fromJson(new String(bytes, StandardCharsets.UTF_8), clazz);
 	}
 
+	public static <T> T unpackGsonWithJson5(Path zip, String path, Class<T> clazz) throws IOException {
+		final byte[] bytes = unpack(zip, path);
+		final var reader = new GsonReader(JsonReader.json5(new String(bytes, StandardCharsets.UTF_8)));
+		return LoomGradlePlugin.GSON.fromJson(reader, clazz);
+	}
+
 	@Nullable
 	public static <T> T unpackGsonNullable(Path zip, String path, Class<T> clazz) throws IOException {
 		try {
@@ -121,6 +129,15 @@ public class ZipUtils {
 		}
 	}
 
+	@Nullable
+	public static <T> T unpackGsonWithJson5Nullable(Path zip, String path, Class<T> clazz) throws IOException {
+		try {
+			return unpackGsonWithJson5(zip, path, clazz);
+		} catch (NoSuchFileException e) {
+			return null;
+		}
+	}
+
 	public static <T> T unpackJson(Path zip, String path, Class<T> clazz) throws IOException {
 		final byte[] bytes = unpack(zip, path);
 		return LoomGradlePlugin.GSON.fromJson(new String(bytes, StandardCharsets.UTF_8), clazz);
@@ -256,6 +273,20 @@ public class ZipUtils {
 		return replacedCount;
 	}
 
+	public static void move(Path zip, String path, String newPath) throws IOException {
+		try (FileSystemUtil.Delegate fs = FileSystemUtil.getJarFileSystem(zip, false)) {
+			var fsPath = fs.get().getPath(path);
+			var newFsPath = fs.get().getPath(newPath);
+
+			if (Files.exists(fsPath)) {
+				Files.write(newFsPath, Files.readAllBytes(fsPath), StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING);
+				Files.delete(fsPath);
+			} else {
+				throw new NoSuchFileException(fsPath.toString());
+			}
+		}
+	}
+
 	@FunctionalInterface
 	public interface UnsafeUnaryOperator<T> {
 		T apply(T arg) throws IOException;
diff --git a/src/main/java/net/fabricmc/loom/util/fmj/FabricModJsonFactory.java b/src/main/java/net/fabricmc/loom/util/fmj/FabricModJsonFactory.java
index f987bcfb37371900fb93ca02e17341f0a5aff129..d76d6f7f3d47e70c2cca7954d834f27f3511e095 100644
--- a/src/main/java/net/fabricmc/loom/util/fmj/FabricModJsonFactory.java
+++ b/src/main/java/net/fabricmc/loom/util/fmj/FabricModJsonFactory.java
@@ -53,6 +53,7 @@ import net.fabricmc.loom.util.gradle.SourceSetHelper;
 public final class FabricModJsonFactory {
 	public static final String FABRIC_MOD_JSON = "fabric.mod.json";
 	public static final String QUILT_MOD_JSON = "quilt.mod.json";
+	public static final String QUILT_MOD_JSON5 = "quilt.mod.json5";
 
 	private static final Logger LOGGER = LoggerFactory.getLogger(FabricModJsonFactory.class);
 
@@ -161,11 +162,15 @@ public final class FabricModJsonFactory {
 	}
 
 	public static boolean isModJar(Path input) {
-		return ZipUtils.contains(input, QUILT_MOD_JSON) || ZipUtils.contains(input, FABRIC_MOD_JSON);
+		return ZipUtils.contains(input, QUILT_MOD_JSON5)
+				|| ZipUtils.contains(input, QUILT_MOD_JSON)
+				|| ZipUtils.contains(input, FABRIC_MOD_JSON);
 	}
 
 	public static boolean containsMod(FileSystemUtil.Delegate fs) {
-		return Files.exists(fs.getPath(QUILT_MOD_JSON)) || Files.exists(fs.getPath(FABRIC_MOD_JSON));
+		return Files.exists(fs.getPath(QUILT_MOD_JSON5))
+				|| Files.exists(fs.getPath(QUILT_MOD_JSON))
+				|| Files.exists(fs.getPath(FABRIC_MOD_JSON));
 	}
 
 	public static boolean isQuiltMod(Path jar) {
@@ -180,7 +185,25 @@ public final class FabricModJsonFactory {
 		}
 	}
 
+	public static boolean isQmj5QuiltMod(Path jar) {
+		try {
+			return ZipUtils.contains(jar, QUILT_MOD_JSON5);
+		} catch (UncheckedIOException e) {
+			if (e.getCause() instanceof NoSuchFileException) {
+				return false;
+			} else {
+				throw e;
+			}
+		}
+	}
+
 	public static String getMetadataPath(Path jar) {
-		return isQuiltMod(jar) ? QUILT_MOD_JSON : FABRIC_MOD_JSON;
+		if (isQmj5QuiltMod(jar)) {
+			return QUILT_MOD_JSON5;
+		} else if (isQuiltMod(jar)) {
+			return QUILT_MOD_JSON;
+		}
+
+		return FABRIC_MOD_JSON;
 	}
 }
diff --git a/src/main/java/net/fabricmc/loom/util/metadata/ModJsonFactory.java b/src/main/java/net/fabricmc/loom/util/metadata/ModJsonFactory.java
index 759f81ed4f5d102f2416c9590591ab7e757f5fef..4f233952da02ec1c779a4aa860d8319b87266fce 100644
--- a/src/main/java/net/fabricmc/loom/util/metadata/ModJsonFactory.java
+++ b/src/main/java/net/fabricmc/loom/util/metadata/ModJsonFactory.java
@@ -40,7 +40,9 @@ import net.fabricmc.loom.util.qmj.QuiltModJsonFactory;
 
 public class ModJsonFactory {
 	public static ModJson createFromZip(Path zipPath) {
-		if (FabricModJsonFactory.isQuiltMod(zipPath)) {
+		if (FabricModJsonFactory.isQmj5QuiltMod(zipPath)) {
+			return QuiltModJsonFactory.createFromJson5Zip(zipPath);
+		} else if (FabricModJsonFactory.isQuiltMod(zipPath)) {
 			return QuiltModJsonFactory.createFromZip(zipPath);
 		} else {
 			return FabricModJsonFactory.createFromZip0(zipPath);
@@ -48,7 +50,9 @@ public class ModJsonFactory {
 	}
 
 	public static ModJson createFromZipNullable(Path zipPath) {
-		if (FabricModJsonFactory.isQuiltMod(zipPath)) {
+		if (FabricModJsonFactory.isQmj5QuiltMod(zipPath)) {
+			return QuiltModJsonFactory.createFromJson5ZipNullable(zipPath);
+		} else if (FabricModJsonFactory.isQuiltMod(zipPath)) {
 			return QuiltModJsonFactory.createFromZipNullable(zipPath);
 		} else {
 			return FabricModJsonFactory.createFromZipNullable0(zipPath);
@@ -56,7 +60,9 @@ public class ModJsonFactory {
 	}
 
 	public static Optional<? extends ModJson> createFromZipOptional(Path zipPath) {
-		if (FabricModJsonFactory.isQuiltMod(zipPath)) {
+		if (FabricModJsonFactory.isQmj5QuiltMod(zipPath)) {
+			return QuiltModJsonFactory.createFromJson5ZipOptional(zipPath);
+		} else if (FabricModJsonFactory.isQuiltMod(zipPath)) {
 			return QuiltModJsonFactory.createFromZipOptional(zipPath);
 		} else {
 			return FabricModJsonFactory.createFromZipOptional0(zipPath);
@@ -68,9 +74,15 @@ public class ModJsonFactory {
 		File file = SourceSetHelper.findFirstFileInResource(FabricModJsonFactory.QUILT_MOD_JSON, project, sourceSets);
 
 		if (file != null) {
-			return QuiltModJsonFactory.createFromSourceSetsNullable(project, sourceSets);
+			return QuiltModJsonFactory.createFromJson5SourceSetsNullable(project, sourceSets);
 		} else {
-			return FabricModJsonFactory.createFromSourceSetsNullable0(project, sourceSets);
+			File qmjFile = SourceSetHelper.findFirstFileInResource(FabricModJsonFactory.QUILT_MOD_JSON, project, sourceSets);
+
+			if (qmjFile != null) {
+				return QuiltModJsonFactory.createFromSourceSetsNullable(project, sourceSets);
+			}
 		}
+
+		return FabricModJsonFactory.createFromSourceSetsNullable0(project, sourceSets);
 	}
 }
diff --git a/src/main/java/net/fabricmc/loom/util/qmj/QuiltModJsonFactory.java b/src/main/java/net/fabricmc/loom/util/qmj/QuiltModJsonFactory.java
index 8f05556168149e2f0e86c5d082db250929f7dbfa..189794ef392ab85ae4a30d92e0210ce15cd69b52 100644
--- a/src/main/java/net/fabricmc/loom/util/qmj/QuiltModJsonFactory.java
+++ b/src/main/java/net/fabricmc/loom/util/qmj/QuiltModJsonFactory.java
@@ -41,6 +41,8 @@ import org.gradle.api.Project;
 import org.gradle.api.tasks.SourceSet;
 import org.jetbrains.annotations.VisibleForTesting;
 import org.jspecify.annotations.Nullable;
+import org.quiltmc.parsers.json.JsonReader;
+import org.quiltmc.parsers.json.gson.GsonReader;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -51,6 +53,7 @@ import net.fabricmc.loom.util.fmj.FabricModJsonSource;
 
 public final class QuiltModJsonFactory {
 	private static final String QUILT_MOD_JSON = "quilt.mod.json";
+	private static final String QUILT_MOD_JSON5 = "quilt.mod.json5";
 	private static final Logger LOGGER = LoggerFactory.getLogger(QuiltModJsonFactory.class);
 	private QuiltModJsonFactory() {
 	}
@@ -74,7 +77,15 @@ public final class QuiltModJsonFactory {
 		try {
 			return create(ZipUtils.unpackGson(zipPath, QUILT_MOD_JSON, JsonObject.class), new FabricModJsonSource.ZipSource(zipPath));
 		} catch (IOException e) {
-			throw new UncheckedIOException("Failed to read fabric.mod.json file in zip: " + zipPath, e);
+			throw new UncheckedIOException("Failed to read quilt.mod.json file in zip: " + zipPath, e);
+		}
+	}
+
+	public static QuiltModJson createFromJson5Zip(Path zipPath) {
+		try {
+			return create(ZipUtils.unpackGsonWithJson5(zipPath, QUILT_MOD_JSON5, JsonObject.class), new FabricModJsonSource.ZipSource(zipPath));
+		} catch (IOException e) {
+			throw new UncheckedIOException("Failed to read quilt.mod.json5 file in zip: " + zipPath, e);
 		}
 	}
 
@@ -95,16 +106,29 @@ public final class QuiltModJsonFactory {
 		return create(jsonObject, new FabricModJsonSource.ZipSource(zipPath));
 	}
 
+	@Nullable
+	public static QuiltModJson createFromJson5ZipNullable(Path zipPath) {
+		JsonObject jsonObject;
+
+		try {
+			jsonObject = ZipUtils.unpackGsonWithJson5Nullable(zipPath, QUILT_MOD_JSON5, JsonObject.class);
+		} catch (IOException e) {
+			throw new UncheckedIOException("Failed to read zip: " + zipPath, e);
+		}
+
+		if (jsonObject == null) {
+			return null;
+		}
+
+		return create(jsonObject, new FabricModJsonSource.ZipSource(zipPath));
+	}
+
 	public static Optional<QuiltModJson> createFromZipOptional(Path zipPath) {
 		return Optional.ofNullable(createFromZipNullable(zipPath));
 	}
 
-	public static QuiltModJson createFromDirectory(Path directory) throws IOException {
-		final Path path = directory.resolve(QUILT_MOD_JSON);
-
-		try (Reader reader = Files.newBufferedReader(path, StandardCharsets.UTF_8)) {
-			return create(LoomGradlePlugin.GSON.fromJson(reader, JsonObject.class), new FabricModJsonSource.DirectorySource(directory));
-		}
+	public static Optional<QuiltModJson> createFromJson5ZipOptional(Path zipPath) {
+		return Optional.ofNullable(createFromJson5ZipNullable(zipPath));
 	}
 
 	@Nullable
@@ -124,4 +148,22 @@ public final class QuiltModJsonFactory {
 			throw new UncheckedIOException("Failed to read " + file.getAbsolutePath(), e);
 		}
 	}
+
+	@Nullable
+	public static QuiltModJson createFromJson5SourceSetsNullable(Project project, SourceSet... sourceSets) throws IOException {
+		final File file = SourceSetHelper.findFirstFileInResource(QUILT_MOD_JSON5, project, sourceSets);
+
+		if (file == null) {
+			return null;
+		}
+
+		try (var reader = JsonReader.json5(Files.newBufferedReader(file.toPath(), StandardCharsets.UTF_8))) {
+			return create(LoomGradlePlugin.GSON.fromJson(new GsonReader(reader), JsonObject.class), new FabricModJsonSource.SourceSetSource(project, sourceSets));
+		} catch (JsonSyntaxException e) {
+			LOGGER.warn("Failed to parse mod json: {}", file.getAbsolutePath());
+			return null;
+		} catch (IOException e) {
+			throw new UncheckedIOException("Failed to read " + file.getAbsolutePath(), e);
+		}
+	}
 }
