From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Glitch <glitch.g3431@gmail.com>
Date: Thu, 9 Jun 2022 23:40:54 -0500
Subject: [PATCH] Exclude Fabric dependencies replaced by Quilt


diff --git a/src/main/java/net/fabricmc/loom/configuration/InstallerData.java b/src/main/java/net/fabricmc/loom/configuration/InstallerData.java
index 9496320b9c7f7c9374f208f5445ceb174a7837f2..4083a9cf3dc8c655743a9c1c58813b681a0aed1f 100644
--- a/src/main/java/net/fabricmc/loom/configuration/InstallerData.java
+++ b/src/main/java/net/fabricmc/loom/configuration/InstallerData.java
@@ -35,7 +35,7 @@ import net.fabricmc.loom.LoomRepositoryPlugin;
 import net.fabricmc.loom.configuration.ide.idea.IdeaUtils;
 import net.fabricmc.loom.util.Constants;
 
-public record InstallerData(String version, JsonObject installerJson) {
+public record InstallerData(String version, JsonObject installerJson, boolean isQuilt) {
 	public void applyToProject(Project project) {
 		LoomGradleExtension extension = LoomGradleExtension.get(project);
 
diff --git a/src/main/java/net/fabricmc/loom/configuration/mods/ArtifactMetadata.java b/src/main/java/net/fabricmc/loom/configuration/mods/ArtifactMetadata.java
index fc3d6484b1801197c93c8f4c77efea10016eb249..536f13bce74690e3a207bd33ec29853ab0480f23 100644
--- a/src/main/java/net/fabricmc/loom/configuration/mods/ArtifactMetadata.java
+++ b/src/main/java/net/fabricmc/loom/configuration/mods/ArtifactMetadata.java
@@ -68,14 +68,16 @@ public record ArtifactMetadata(boolean isFabricMod, RemapRequirements remapRequi
 			}
 
 			Path installerPath = fs.getPath(QUILT_INSTALLER_PATH);
+			boolean isQuilt = true;
 
 			if (!Files.exists(installerPath)) {
-				installerPath = fs.getPath(FABRIC_INSTALLER_PATH);;
+				installerPath = fs.getPath(FABRIC_INSTALLER_PATH);
+				isQuilt = false;
 			}
 
 			if (isFabricMod && Files.exists(installerPath)) {
 				final JsonObject jsonObject = LoomGradlePlugin.GSON.fromJson(Files.readString(installerPath, StandardCharsets.UTF_8), JsonObject.class);
-				installerData = new InstallerData(artifact.version(), jsonObject);
+				installerData = new InstallerData(artifact.version(), jsonObject, isQuilt);
 			}
 		}
 
diff --git a/src/main/java/net/fabricmc/loom/configuration/mods/ModConfigurationRemapper.java b/src/main/java/net/fabricmc/loom/configuration/mods/ModConfigurationRemapper.java
index 7ff0fbc78dc242a82a925ca72dd85168116aa115..cdb27fca02bae7b2f26864a90b69c812876f202d 100644
--- a/src/main/java/net/fabricmc/loom/configuration/mods/ModConfigurationRemapper.java
+++ b/src/main/java/net/fabricmc/loom/configuration/mods/ModConfigurationRemapper.java
@@ -31,12 +31,20 @@ import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Supplier;
 
 import com.google.common.collect.ImmutableMap;
+
+import net.fabricmc.loom.configuration.InstallerData;
+
+import net.fabricmc.loom.configuration.mods.dependency.ModDependencyFactory;
+
 import org.gradle.api.Project;
 import org.gradle.api.artifacts.Configuration;
 import org.gradle.api.artifacts.FileCollectionDependency;
@@ -59,7 +67,6 @@ import net.fabricmc.loom.LoomGradleExtension;
 import net.fabricmc.loom.api.RemapConfigurationSettings;
 import net.fabricmc.loom.configuration.RemapConfigurations;
 import net.fabricmc.loom.configuration.mods.dependency.ModDependency;
-import net.fabricmc.loom.configuration.mods.dependency.ModDependencyFactory;
 import net.fabricmc.loom.configuration.providers.minecraft.MinecraftSourceSets;
 import net.fabricmc.loom.util.Checksum;
 import net.fabricmc.loom.util.Constants;
@@ -121,19 +128,19 @@ public class ModConfigurationRemapper {
 		}
 
 		// Round 1: Discovery
-		// Go through all the configs to find artifacts to remap and
-		// the installer data. The installer data has to be added before
+		// Go through all the configs and find the installer data and QSL.
+		// The installer data has to be added before
 		// any mods are remapped since remapping needs the dependencies provided by that data.
+		// The presence of Quilt Loader/QSL will be used to exclude the Fabric equivalents if needed.
+
 		final Map<Configuration, List<ModDependency>> dependenciesBySourceConfig = new HashMap<>();
+		final Map<Configuration, List<ArtifactRef>> artifactsBySourceConfig = new HashMap<>();
+		final Set<Configuration> sourceConfigsWithQsl = new HashSet<>();
+		AtomicReference<InstallerData> installer = new AtomicReference<>();
 		configsToRemap.forEach((sourceConfig, remappedConfig) -> {
-			/*
-			sourceConfig - The source configuration where the intermediary named artifacts come from. i.e "modApi"
-			remappedConfig - The target configuration where the remapped artifacts go
-			 */
-			final Configuration clientRemappedConfig = clientConfigsToRemap.get(sourceConfig);
-			final List<ModDependency> modDependencies = new ArrayList<>();
+			List<ArtifactRef> artifacts = resolveArtifacts(project, sourceConfig);
 
-			for (ArtifactRef artifact : resolveArtifacts(project, sourceConfig)) {
+			for (ArtifactRef artifact : artifacts) {
 				final ArtifactMetadata artifactMetadata;
 
 				try {
@@ -143,11 +150,13 @@ public class ModConfigurationRemapper {
 				}
 
 				if (artifactMetadata.installerData() != null) {
-					if (extension.getInstallerData() != null) {
-						project.getLogger().info("Found another installer JSON in ({}), ignoring", artifact.path());
+					if (installer.get() == null) {
+						installer.set(artifactMetadata.installerData());
+					} else if (!installer.get().isQuilt() && artifactMetadata.installerData().isQuilt()) {
+						// Prefer Quilt over Fabric loader
+						installer.set(artifactMetadata.installerData());
 					} else {
-						project.getLogger().info("Applying installer data from {}", artifact.path());
-						artifactMetadata.installerData().applyToProject(project);
+						project.getLogger().info("Found another installer JSON in ({}), ignoring", artifact.path());
 					}
 				}
 
@@ -158,6 +167,49 @@ public class ModConfigurationRemapper {
 					continue;
 				}
 
+				if (artifact.group().startsWith("org.quiltmc.qsl")) {
+					sourceConfigsWithQsl.add(sourceConfig);
+				}
+			}
+
+			artifactsBySourceConfig.put(sourceConfig, artifacts);
+		});
+
+		boolean quiltLoader;
+
+		// Apply the installer.
+		// We do this outside Round 1, unlike Fabric Loom, because we need to prioritize Quilt Loader over Fabric Loader.
+		if (installer.get() != null) {
+			project.getLogger().info("Applying installer data");
+
+            if (installer.get().isQuilt()) {
+                quiltLoader = true;
+            } else {
+				quiltLoader = false;
+				project.getLogger().lifecycle("Warning: Quilt Loader not detected. Using Fabric Loader.");
+            }
+
+            installer.get().applyToProject(project);
+		} else {
+			quiltLoader = false;
+		}
+
+		// Round 1.5: Proposal
+		// Propose everything to be remapped, excluding anything we want to hide.
+		configsToRemap.forEach((sourceConfig, remappedConfig) -> {
+			/*
+			sourceConfig - The source configuration where the intermediary named artifacts come from. i.e "modApi"
+			remappedConfig - The target configuration where the remapped artifacts go
+			 */
+			final Configuration clientRemappedConfig = clientConfigsToRemap.get(sourceConfig);
+			final List<ModDependency> modDependencies = new ArrayList<>();
+
+			for (ArtifactRef artifact : artifactsBySourceConfig.get(sourceConfig)) {
+				if (quiltLoader && artifact.group().equals("net.fabricmc") && artifact.name().equals("fabric-loader")) {
+					continue;
+				} else if (sourceConfigsWithQsl.contains(sourceConfig) && artifact.group().equals("net.fabricmc.fabric-api")) {
+					continue;
+				}
 				final ModDependency modDependency = ModDependencyFactory.create(artifact, remappedConfig, clientRemappedConfig, mappingsSuffix, project);
 				scheduleSourcesRemapping(project, sourceRemapper, modDependency);
 				modDependencies.add(modDependency);
diff --git a/src/test/groovy/net/fabricmc/loom/test/integration/MavenProjectTest.groovy b/src/test/groovy/net/fabricmc/loom/test/integration/MavenProjectTest.groovy
index 1a546c22e8ee8d4c56f9dd346a89b319402bbb2c..b56f85b7a4674b652b5bfc2c9ddf8d2daeec3a1d 100644
--- a/src/test/groovy/net/fabricmc/loom/test/integration/MavenProjectTest.groovy
+++ b/src/test/groovy/net/fabricmc/loom/test/integration/MavenProjectTest.groovy
@@ -46,6 +46,7 @@ class MavenProjectTest extends Specification implements MockMavenServerTrait, Gr
 	def "publish lib #version #gradleVersion"() {
 		setup:
 		setProperty('loom.test.version', version)
+		setProperty('loom.excludeFabricReplacedDependencies', 'false')
 		def gradle = gradleProject(project: "mavenLibrary", version: gradleVersion, sharedFiles: true)
 
 		when:
@@ -73,6 +74,7 @@ class MavenProjectTest extends Specification implements MockMavenServerTrait, Gr
 	def "resolve #version #gradleVersion"() {
 		given:
 		setProperty('loom.test.resolve', "com.example:fabric-example-lib:${version}")
+		setProperty('loom.excludeFabricReplacedDependencies', 'false')
 		def gradle = gradleProject(project: "maven", version: gradleVersion, sharedFiles: true)
 
 		when:
diff --git a/src/test/groovy/net/fabricmc/loom/test/integration/MultiProjectTest.groovy b/src/test/groovy/net/fabricmc/loom/test/integration/MultiProjectTest.groovy
index da86621384189424e7a03ba0a9d6e3dcddc64e51..1f102b9348d343e6333d15dbfe4ba89c12042887 100644
--- a/src/test/groovy/net/fabricmc/loom/test/integration/MultiProjectTest.groovy
+++ b/src/test/groovy/net/fabricmc/loom/test/integration/MultiProjectTest.groovy
@@ -29,6 +29,7 @@ import spock.lang.Unroll
 
 import net.fabricmc.loom.test.util.GradleProjectTestTrait
 
+import static java.lang.System.setProperty
 import static net.fabricmc.loom.test.LoomTestConstants.STANDARD_TEST_VERSIONS
 import static org.gradle.testkit.runner.TaskOutcome.SUCCESS
 
@@ -36,6 +37,7 @@ class MultiProjectTest extends Specification implements GradleProjectTestTrait {
 	@Unroll
 	def "build (gradle #version)"() {
 		setup:
+		setProperty('loom.excludeFabricReplacedDependencies', 'false')
 		def gradle = gradleProject(project: "multiproject", version: version)
 
 		when:
