From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Glitch <glitch.g3431@gmail.com>
Date: Thu, 9 Jun 2022 23:40:54 -0500
Subject: [PATCH] Exclude Fabric dependencies replaced by Quilt


diff --git a/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java b/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java
index 0a4c7c9036810f6ef8ac2903d49638975231bee0..80d3f734c9722233f0edb8af67f7a3fed01b8da2 100644
--- a/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java
+++ b/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java
@@ -24,6 +24,7 @@
 
 package net.fabricmc.loom;
 
+import java.util.Map;
 import java.util.Objects;
 
 import com.fasterxml.jackson.databind.DeserializationFeature;
@@ -83,5 +84,14 @@ public class LoomGradlePlugin implements BootstrappedPlugin {
 		IdeaConfiguration.setup(project);
 
 		project.apply(ImmutableMap.of("plugin", "io.github.juuxel.loom-quiltflower"));
+
+		project.afterEvaluate(proj -> {
+			if (LoomGradleExtension.get(project).excludeFabricReplacedDependencies().get()) {
+				project.getConfigurations().all(conf -> {
+					conf.exclude(Map.of("group", "net.fabricmc", "module", "fabric-loader"));
+					conf.exclude(Map.of("group", "net.fabricmc.fabric-api"));
+				});
+			}
+		});
 	}
 }
diff --git a/src/main/java/net/fabricmc/loom/api/LoomGradleExtensionAPI.java b/src/main/java/net/fabricmc/loom/api/LoomGradleExtensionAPI.java
index 45587d96394ba1e4d9dffb790d80a76b4b3f39f8..731b9196f732623bbdaf22005fe7688611c9c47e 100644
--- a/src/main/java/net/fabricmc/loom/api/LoomGradleExtensionAPI.java
+++ b/src/main/java/net/fabricmc/loom/api/LoomGradleExtensionAPI.java
@@ -218,4 +218,10 @@ public interface LoomGradleExtensionAPI {
 	Property<Boolean> getRuntimeOnlyLog4j();
 
 	Property<Boolean> getSplitModDependencies();
+
+	/**
+	 * Whether or not to exclude dependencies that have Fabric replacements, like Fabric Loader and Fabric API.
+	 * Defaults to true.
+	 */
+	Property<Boolean> excludeFabricReplacedDependencies();
 }
diff --git a/src/main/java/net/fabricmc/loom/extension/LoomGradleExtensionApiImpl.java b/src/main/java/net/fabricmc/loom/extension/LoomGradleExtensionApiImpl.java
index 17e54ec3e5aa1f9f7f8e1c85a6a57b2e654923fe..fdd9c5aa4538df01eb78640163a00e8fefcd94c1 100644
--- a/src/main/java/net/fabricmc/loom/extension/LoomGradleExtensionApiImpl.java
+++ b/src/main/java/net/fabricmc/loom/extension/LoomGradleExtensionApiImpl.java
@@ -77,6 +77,7 @@ public abstract class LoomGradleExtensionApiImpl implements LoomGradleExtensionA
 	protected final Property<IntermediateMappingsProvider> intermediateMappingsProvider;
 	private final Property<Boolean> runtimeOnlyLog4j;
 	private final Property<Boolean> splitModDependencies;
+	private final Property<Boolean> replaceFabric;
 	private final Property<MinecraftJarConfiguration> minecraftJarConfiguration;
 	private final Property<Boolean> splitEnvironmentalSourceSet;
 	private final InterfaceInjectionExtensionAPI interfaceInjectionExtension;
@@ -135,6 +136,8 @@ public abstract class LoomGradleExtensionApiImpl implements LoomGradleExtensionA
 		this.splitModDependencies = project.getObjects().property(Boolean.class).convention(true);
 		this.splitModDependencies.finalizeValueOnRead();
 
+		this.replaceFabric = project.getObjects().property(Boolean.class).convention(true);
+
 		this.interfaceInjectionExtension = project.getObjects().newInstance(InterfaceInjectionExtensionAPI.class);
 		this.interfaceInjectionExtension.getIsEnabled().convention(true);
 
@@ -368,6 +371,10 @@ public abstract class LoomGradleExtensionApiImpl implements LoomGradleExtensionA
 		RemapConfigurations.setupForSourceSet(getProject(), sourceSet);
 	}
 
+	public Property<Boolean> excludeFabricReplacedDependencies() {
+		return replaceFabric;
+	}
+
 	// This is here to ensure that LoomGradleExtensionApiImpl compiles without any unimplemented methods
 	private final class EnsureCompile extends LoomGradleExtensionApiImpl {
 		private EnsureCompile() {
