From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kneelawk <kneelawk@gmail.com>
Date: Sun, 1 May 2022 19:11:28 -0700
Subject: [PATCH] Add metadata priority resolution logic


diff --git a/src/main/java/net/fabricmc/loom/util/ModUtils.java b/src/main/java/net/fabricmc/loom/util/ModUtils.java
index 6e02f75be166df73665483ee3f925e1d56648cde..7109ad9219c00966f000a68acdac41fb8a464d99 100644
--- a/src/main/java/net/fabricmc/loom/util/ModUtils.java
+++ b/src/main/java/net/fabricmc/loom/util/ModUtils.java
@@ -35,6 +35,11 @@ import net.fabricmc.loom.api.LoomGradleExtensionAPI;
 import net.fabricmc.loom.configuration.ModMetadataHelper;
 
 public final class ModUtils {
+	/**
+	 * Priority of different metadata types. Later metadatas have a higher priority.
+	 */
+	private static final List<String> METADATA_PRIORITY = List.of("fabric.mod.json", "quilt.mod.json");
+
 	private ModUtils() {
 	}
 
@@ -47,14 +52,14 @@ public final class ModUtils {
 	}
 
 	/**
-	 * @throws UnsupportedOperationException if the jar file has more than one kind of metadata, or the metadata that is found cannot be read.
+	 * @throws UnsupportedOperationException if the metadata that is found cannot be read.
 	 */
 	public static ModMetadataHelper.Metadata readMetadataFromJar(LoomGradleExtensionAPI ext, File jar) {
 		return ModUtils.readMetadataFromJar(ext.getModMetadataHelpers().get(), jar);
 	}
 
 	/**
-	 * @throws UnsupportedOperationException if the jar file has more than one kind of metadata, or the metadata that is found cannot be read.
+	 * @throws UnsupportedOperationException if the metadata that is found cannot be read.
 	 */
 	public static ModMetadataHelper.Metadata readMetadataFromJar(Map<String, ModMetadataHelper> helpers, File jar) {
 		try (var zip = new ZipFile(jar)) {
@@ -67,11 +72,21 @@ public final class ModUtils {
 				return null;
 			}
 
+			// Picks the metadata with the highest priority.
+			String fileName = entries.get(0);
+
 			if (entries.size() > 1) {
-				throw new UnsupportedOperationException("Cannot read jars with more than one kind of metadata.");
-			}
+				int curPriority = METADATA_PRIORITY.indexOf(fileName);
 
-			String fileName = entries.get(0);
+				for (String entry : entries) {
+					int newPriority = METADATA_PRIORITY.indexOf(entry);
+
+					if (newPriority > curPriority) {
+						curPriority = newPriority;
+						fileName = entry;
+					}
+				}
+			}
 
 			try (InputStreamReader reader = new InputStreamReader(zip.getInputStream(zip.getEntry(fileName)))) {
 				return helpers.get(fileName).createMetadata(reader);
