From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Glitch <glitch.g3431@gmail.com>
Date: Thu, 9 Jun 2022 23:40:54 -0500
Subject: [PATCH] Exclude Fabric dependencies replaced by Quilt


diff --git a/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java b/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java
index 0a4c7c9036810f6ef8ac2903d49638975231bee0..210a7ea71b754a52b47eb66591acae9cc43d81af 100644
--- a/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java
+++ b/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java
@@ -24,6 +24,7 @@
 
 package net.fabricmc.loom;
 
+import java.util.Map;
 import java.util.Objects;
 
 import com.fasterxml.jackson.databind.DeserializationFeature;
@@ -83,5 +84,12 @@ public class LoomGradlePlugin implements BootstrappedPlugin {
 		IdeaConfiguration.setup(project);
 
 		project.apply(ImmutableMap.of("plugin", "io.github.juuxel.loom-quiltflower"));
+
+		if (!"false".equalsIgnoreCase(System.getProperty("loom.excludeFabricReplacedDependencies"))) {
+			project.getConfigurations().all(a -> {
+				a.exclude(Map.of("group", "net.fabricmc", "module", "fabric-loader"));
+				a.exclude(Map.of("group", "net.fabricmc.fabric-api"));
+			});
+		}
 	}
 }
diff --git a/src/test/groovy/net/fabricmc/loom/test/integration/MavenProjectTest.groovy b/src/test/groovy/net/fabricmc/loom/test/integration/MavenProjectTest.groovy
index 1a546c22e8ee8d4c56f9dd346a89b319402bbb2c..b56f85b7a4674b652b5bfc2c9ddf8d2daeec3a1d 100644
--- a/src/test/groovy/net/fabricmc/loom/test/integration/MavenProjectTest.groovy
+++ b/src/test/groovy/net/fabricmc/loom/test/integration/MavenProjectTest.groovy
@@ -46,6 +46,7 @@ class MavenProjectTest extends Specification implements MockMavenServerTrait, Gr
 	def "publish lib #version #gradleVersion"() {
 		setup:
 		setProperty('loom.test.version', version)
+		setProperty('loom.excludeFabricReplacedDependencies', 'false')
 		def gradle = gradleProject(project: "mavenLibrary", version: gradleVersion, sharedFiles: true)
 
 		when:
@@ -73,6 +74,7 @@ class MavenProjectTest extends Specification implements MockMavenServerTrait, Gr
 	def "resolve #version #gradleVersion"() {
 		given:
 		setProperty('loom.test.resolve', "com.example:fabric-example-lib:${version}")
+		setProperty('loom.excludeFabricReplacedDependencies', 'false')
 		def gradle = gradleProject(project: "maven", version: gradleVersion, sharedFiles: true)
 
 		when:
diff --git a/src/test/groovy/net/fabricmc/loom/test/integration/MultiProjectTest.groovy b/src/test/groovy/net/fabricmc/loom/test/integration/MultiProjectTest.groovy
index da86621384189424e7a03ba0a9d6e3dcddc64e51..1f102b9348d343e6333d15dbfe4ba89c12042887 100644
--- a/src/test/groovy/net/fabricmc/loom/test/integration/MultiProjectTest.groovy
+++ b/src/test/groovy/net/fabricmc/loom/test/integration/MultiProjectTest.groovy
@@ -29,6 +29,7 @@ import spock.lang.Unroll
 
 import net.fabricmc.loom.test.util.GradleProjectTestTrait
 
+import static java.lang.System.setProperty
 import static net.fabricmc.loom.test.LoomTestConstants.STANDARD_TEST_VERSIONS
 import static org.gradle.testkit.runner.TaskOutcome.SUCCESS
 
@@ -36,6 +37,7 @@ class MultiProjectTest extends Specification implements GradleProjectTestTrait {
 	@Unroll
 	def "build (gradle #version)"() {
 		setup:
+		setProperty('loom.excludeFabricReplacedDependencies', 'false')
 		def gradle = gradleProject(project: "multiproject", version: version)
 
 		when:
