From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Glitch <glitch.g3431@gmail.com>
Date: Thu, 9 Jun 2022 23:40:54 -0500
Subject: [PATCH] Exclude Fabric dependencies replaced by Quilt


diff --git a/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java b/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java
index 0a4c7c9036810f6ef8ac2903d49638975231bee0..210a7ea71b754a52b47eb66591acae9cc43d81af 100644
--- a/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java
+++ b/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java
@@ -24,6 +24,7 @@
 
 package net.fabricmc.loom;
 
+import java.util.Map;
 import java.util.Objects;
 
 import com.fasterxml.jackson.databind.DeserializationFeature;
@@ -83,5 +84,12 @@ public class LoomGradlePlugin implements BootstrappedPlugin {
 		IdeaConfiguration.setup(project);
 
 		project.apply(ImmutableMap.of("plugin", "io.github.juuxel.loom-quiltflower"));
+
+		if (!"false".equalsIgnoreCase(System.getProperty("loom.excludeFabricReplacedDependencies"))) {
+			project.getConfigurations().all(a -> {
+				a.exclude(Map.of("group", "net.fabricmc", "module", "fabric-loader"));
+				a.exclude(Map.of("group", "net.fabricmc.fabric-api"));
+			});
+		}
 	}
 }
diff --git a/src/test/groovy/net/fabricmc/loom/test/integration/MavenProjectTest.groovy b/src/test/groovy/net/fabricmc/loom/test/integration/MavenProjectTest.groovy
index cda61b1b4a32bb74148e4fea20ea1ca7041424df..bb15f0ae2adc788383a306652f8546223d75de7c 100644
--- a/src/test/groovy/net/fabricmc/loom/test/integration/MavenProjectTest.groovy
+++ b/src/test/groovy/net/fabricmc/loom/test/integration/MavenProjectTest.groovy
@@ -45,6 +45,7 @@ class MavenProjectTest extends Specification implements MockMavenServerTrait, Gr
 	def "publish lib #version #gradleVersion"() {
 		setup:
 			setProperty('loom.test.version', version)
+			setProperty('loom.excludeFabricReplacedDependencies', 'false')
 			def gradle = gradleProject(project: "mavenLibrary", version: gradleVersion, sharedFiles: true)
 
 		when:
@@ -72,6 +73,7 @@ class MavenProjectTest extends Specification implements MockMavenServerTrait, Gr
 	def "resolve #version #gradleVersion"() {
 		given:
 			setProperty('loom.test.resolve', "com.example:fabric-example-lib:${version}")
+			setProperty('loom.excludeFabricReplacedDependencies', 'false')
 			def gradle = gradleProject(project: "maven", version: gradleVersion, sharedFiles: true)
 
 		when:
diff --git a/src/test/groovy/net/fabricmc/loom/test/integration/MultiProjectTest.groovy b/src/test/groovy/net/fabricmc/loom/test/integration/MultiProjectTest.groovy
index 3984e0042fa3ade628992920270714e46b2db28f..0762ef5c0b7c1f00814ca5f9b7a44c1b6c793dfa 100644
--- a/src/test/groovy/net/fabricmc/loom/test/integration/MultiProjectTest.groovy
+++ b/src/test/groovy/net/fabricmc/loom/test/integration/MultiProjectTest.groovy
@@ -28,6 +28,7 @@ import net.fabricmc.loom.test.util.GradleProjectTestTrait
 import spock.lang.Specification
 import spock.lang.Unroll
 
+import static java.lang.System.setProperty
 import static net.fabricmc.loom.test.LoomTestConstants.STANDARD_TEST_VERSIONS
 import static org.gradle.testkit.runner.TaskOutcome.SUCCESS
 
@@ -35,6 +36,7 @@ class MultiProjectTest extends Specification implements GradleProjectTestTrait {
 	@Unroll
 	def "build (gradle #version)"() {
 		setup:
+			setProperty('loom.excludeFabricReplacedDependencies', 'false')
 			def gradle = gradleProject(project: "multiproject", version: version)
 
 		when:
