From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Glitch <glitch.g3431@gmail.com>
Date: Thu, 9 Jun 2022 23:40:54 -0500
Subject: [PATCH] Exclude Fabric dependencies replaced by Quilt


diff --git a/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java b/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java
index 969db77e44263f58bd0fa40b6e08e0b7be3efd8f..fcb8e99d9bfc1dc2838013b11b0bbde5464122e3 100644
--- a/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java
+++ b/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java
@@ -24,6 +24,7 @@
 
 package net.fabricmc.loom;
 
+import java.util.Map;
 import java.util.Objects;
 
 import com.fasterxml.jackson.databind.DeserializationFeature;
@@ -84,5 +85,12 @@ public class LoomGradlePlugin implements BootstrappedPlugin {
 
 		project.apply(ImmutableMap.of("plugin", "org.quiltmc.quilt-mappings-on-loom"));
 		project.apply(ImmutableMap.of("plugin", "io.github.juuxel.loom-quiltflower"));
+
+		if (LoomGradleExtension.get(project).excludeFabricReplacedDependencies().get()) {
+			project.getConfigurations().all(a -> {
+				a.exclude(Map.of("group", "net.fabricmc", "module", "fabric-loader"));
+				a.exclude(Map.of("group", "net.fabricmc.fabric-api"));
+			});
+		}
 	}
 }
diff --git a/src/main/java/net/fabricmc/loom/api/LoomGradleExtensionAPI.java b/src/main/java/net/fabricmc/loom/api/LoomGradleExtensionAPI.java
index 18cb0f84a070c425f151864b0ecb04a2fa14b107..aa7e0178aec8bedf9ea0ea04ae146a26250447f3 100644
--- a/src/main/java/net/fabricmc/loom/api/LoomGradleExtensionAPI.java
+++ b/src/main/java/net/fabricmc/loom/api/LoomGradleExtensionAPI.java
@@ -203,4 +203,10 @@ public interface LoomGradleExtensionAPI {
 	Property<Boolean> getRuntimeOnlyLog4j();
 
 	Property<Boolean> getSplitModDependencies();
+
+	/**
+	 * Whether or not to exclude dependencies that have Fabric replacements, like Fabric Loader and Fabric API.
+	 * Defaults to true.
+	 */
+	Property<Boolean> excludeFabricReplacedDependencies();
 }
diff --git a/src/main/java/net/fabricmc/loom/extension/LoomGradleExtensionApiImpl.java b/src/main/java/net/fabricmc/loom/extension/LoomGradleExtensionApiImpl.java
index a5e6abf1f0d3e92abd67af3d846bcaf9510070ff..219cd26d1429cfc6e44e1a3b864ede284d5cb446 100644
--- a/src/main/java/net/fabricmc/loom/extension/LoomGradleExtensionApiImpl.java
+++ b/src/main/java/net/fabricmc/loom/extension/LoomGradleExtensionApiImpl.java
@@ -77,6 +77,7 @@ public abstract class LoomGradleExtensionApiImpl implements LoomGradleExtensionA
 	protected final Property<IntermediateMappingsProvider> intermediateMappingsProvider;
 	private final Property<Boolean> runtimeOnlyLog4j;
 	private final Property<Boolean> splitModDependencies;
+	private final Property<Boolean> replaceFabric;
 	private final Property<MinecraftJarConfiguration> minecraftJarConfiguration;
 	private final Property<Boolean> splitEnvironmentalSourceSet;
 	private final InterfaceInjectionExtensionAPI interfaceInjectionExtension;
@@ -132,6 +133,9 @@ public abstract class LoomGradleExtensionApiImpl implements LoomGradleExtensionA
 		this.splitModDependencies = project.getObjects().property(Boolean.class).convention(true);
 		this.splitModDependencies.finalizeValueOnRead();
 
+		this.replaceFabric = project.getObjects().property(Boolean.class).convention(true);
+		this.replaceFabric.finalizeValueOnRead();
+
 		this.interfaceInjectionExtension = project.getObjects().newInstance(InterfaceInjectionExtensionAPI.class);
 
 		this.splitEnvironmentalSourceSet = project.getObjects().property(Boolean.class).convention(false);
@@ -358,6 +362,10 @@ public abstract class LoomGradleExtensionApiImpl implements LoomGradleExtensionA
 		RemapConfigurations.setupForSourceSet(getProject(), sourceSet);
 	}
 
+	public Property<Boolean> excludeFabricReplacedDependencies() {
+		return replaceFabric;
+	}
+
 	// This is here to ensure that LoomGradleExtensionApiImpl compiles without any unimplemented methods
 	private final class EnsureCompile extends LoomGradleExtensionApiImpl {
 		private EnsureCompile() {
