From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tibs <git@tibinonest.me>
Date: Sat, 9 Jul 2022 15:45:07 -0400
Subject: [PATCH] Fix kotlin metadata remapping for inner classes


diff --git a/src/main/kotlin/net/fabricmc/loom/kotlin/remapping/RemappingKmVisitors.kt b/src/main/kotlin/net/fabricmc/loom/kotlin/remapping/RemappingKmVisitors.kt
index 7cb9ed04d72abaeb2104e45418f38b486cac9155..ce8777d032bbb0624219ba431f14f650fdf3233a 100644
--- a/src/main/kotlin/net/fabricmc/loom/kotlin/remapping/RemappingKmVisitors.kt
+++ b/src/main/kotlin/net/fabricmc/loom/kotlin/remapping/RemappingKmVisitors.kt
@@ -138,11 +138,11 @@ class RemappingKmVisitors(private val remapper: Remapper) {
 
     inner class RemappingKmTypeVisitor(delegate: KmTypeVisitor?) : KmTypeVisitor(delegate) {
         override fun visitClass(name: ClassName) {
-            super.visitClass(remapper.map(name))
+            super.visitClass(remapper.map(name.replace('.', '$')).replace('$', '.'))
         }
 
         override fun visitTypeAlias(name: ClassName) {
-            super.visitTypeAlias(remapper.map(name))
+            super.visitTypeAlias(remapper.map(name.replace('.', '$')).replace('$', '.'))
         }
 
         override fun visitAbbreviatedType(flags: Flags): KmTypeVisitor {
