From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nik-Kot <42574700+Nik-Kot@users.noreply.github.com>
Date: Fri, 14 Apr 2023 21:08:03 -0500
Subject: [PATCH] Fix compatibility with fabric-loom 1.1.x by resolving shared
 service manager conflict (renaming the instance)


diff --git a/src/main/java/net/fabricmc/loom/util/service/BuildSharedServiceManager.java b/src/main/java/net/fabricmc/loom/util/service/BuildSharedServiceManager.java
index 6575611dea711ca1bb284a76132110404cdd95cb..b6f0ed542dc183f5782e300c57dce2c3ee3969b2 100644
--- a/src/main/java/net/fabricmc/loom/util/service/BuildSharedServiceManager.java
+++ b/src/main/java/net/fabricmc/loom/util/service/BuildSharedServiceManager.java
@@ -39,7 +39,7 @@ import org.slf4j.LoggerFactory;
 
 public abstract class BuildSharedServiceManager implements BuildService<BuildServiceParameters.None> {
 	private static final Logger LOGGER = LoggerFactory.getLogger(BuildSharedServiceManager.class);
-	private static final String NAME = "loom:sharedServiceManager";
+	private static final String NAME = "quilt-loom:sharedServiceManager";
 
 	private SharedServiceManager sharedServiceManager = new BuildSharedServiceManagerImpl();
 	private final AtomicInteger refCount = new AtomicInteger(0);
